File: avalon-modules/avalon-dataset/avalon-dataset-controller/src/test/java/io/github/lishangbu/avalon/dataset/controller/BerryFlavorControllerTest.java
package io.github.lishangbu.avalon.dataset.controller;

import static org.junit.jupiter.api.Assertions.assertSame;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import io.github.lishangbu.avalon.dataset.entity.BerryFlavor;
import io.github.lishangbu.avalon.dataset.service.BerryFlavorService;
import java.util.List;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;

@ExtendWith(MockitoExtension.class)
class BerryFlavorControllerTest {
    @Mock private BerryFlavorService berryFlavorService;
    @InjectMocks private BerryFlavorController berryFlavorController;

    @Test
    void getBerryFlavorPage_delegatesToService() {
        BerryFlavor berryFlavor = new BerryFlavor();
        Pageable pageable = PageRequest.of(0, 5);
        Page<BerryFlavor> page = new PageImpl<>(List.of(berryFlavor));
        when(berryFlavorService.getPageByCondition(any(BerryFlavor.class), eq(pageable)))
                .thenReturn(page);

        Page<BerryFlavor> result = berryFlavorController.getBerryFlavorPage(pageable, berryFlavor);

        assertSame(page, result);
        verify(berryFlavorService).getPageByCondition(berryFlavor, pageable);
    }

    @Test
    void save_delegatesToService() {
        BerryFlavor berryFlavor = new BerryFlavor();
        when(berryFlavorService.save(berryFlavor)).thenReturn(berryFlavor);

        BerryFlavor result = berryFlavorController.save(berryFlavor);

        assertSame(berryFlavor, result);
        verify(berryFlavorService).save(berryFlavor);
    }

    @Test
    void update_delegatesToService() {
        BerryFlavor berryFlavor = new BerryFlavor();
        when(berryFlavorService.update(berryFlavor)).thenReturn(berryFlavor);

        BerryFlavor result = berryFlavorController.update(berryFlavor);

        assertSame(berryFlavor, result);
        verify(berryFlavorService).update(berryFlavor);
    }

    @Test
    void deleteById_delegatesToService() {
        berryFlavorController.deleteById(1L);
        verify(berryFlavorService).removeById(1L);
    }
}

File: avalon-modules/avalon-dataset/avalon-dataset-controller/src/test/kotlin/io/github/lishangbu/avalon/dataset/controller/BerryFlavorControllerTest.kt
package io.github.lishangbu.avalon.dataset.controller

import io.github.lishangbu.avalon.dataset.entity.dto.BerryFlavorSpecification
import io.github.lishangbu.avalon.dataset.entity.dto.BerryFlavorView
import io.github.lishangbu.avalon.dataset.entity.dto.SaveBerryFlavorInput
import io.github.lishangbu.avalon.dataset.entity.dto.UpdateBerryFlavorInput
import io.github.lishangbu.avalon.dataset.service.BerryFlavorService
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertSame
import org.junit.jupiter.api.Test

class BerryFlavorControllerTest {
    @Test
    fun listBerryFlavors_delegatesToService() {
        val service = FakeBerryFlavorService()
        val controller = BerryFlavorController(service)
        val list = listOf(BerryFlavorView("1", "spicy", "辣"))
        service.listResult = list
        val specification = BerryFlavorSpecification(id = "1", internalName = "spicy", name = "辣")

        val result = controller.listBerryFlavors(specification)

        assertSame(list, result)
        assertEquals("1", service.listCondition!!.id)
    }

    @Test
    fun save_delegatesToService() {
        val service = FakeBerryFlavorService()
        val controller = BerryFlavorController(service)
        val command = SaveBerryFlavorInput("spicy", "辣")
        service.saveResult = BerryFlavorView("1", "spicy", "辣")

        val result = controller.save(command)

        assertSame(service.saveResult, result)
        assertSame(command, service.savedCommand)
    }

    @Test
    fun update_delegatesToService() {
        val service = FakeBerryFlavorService()
        val controller = BerryFlavorController(service)
        val command = UpdateBerryFlavorInput("1", "spicy", "辣")
        service.updateResult = BerryFlavorView("1", "spicy", "辣")

        val result = controller.update(command)

        assertSame(service.updateResult, result)
        assertSame(command, service.updatedCommand)
    }

    @Test
    fun deleteById_delegatesToService() {
        val service = FakeBerryFlavorService()
        val controller = BerryFlavorController(service)

        controller.deleteById(1L)

        assertEquals(1L, service.removedId)
    }

    private class FakeBerryFlavorService : BerryFlavorService {
        var listCondition: BerryFlavorSpecification? = null
        var savedCommand: SaveBerryFlavorInput? = null
        var updatedCommand: UpdateBerryFlavorInput? = null
        var removedId: Long? = null

        var listResult: List<BerryFlavorView> = emptyList()
        lateinit var saveResult: BerryFlavorView
        lateinit var updateResult: BerryFlavorView

        override fun save(command: SaveBerryFlavorInput): BerryFlavorView {
            savedCommand = command
            return saveResult
        }

        override fun update(command: UpdateBerryFlavorInput): BerryFlavorView {
            updatedCommand = command
            return updateResult
        }

        override fun removeById(id: Long) {
            removedId = id
        }

        override fun listByCondition(specification: BerryFlavorSpecification): List<BerryFlavorView> {
            listCondition = specification
            return listResult
        }
    }
}
